Make Text and Tex robust against non-existing text_dir and tex_dir

behackl opened the pull request:
## List of Changes
- When creating Tex and Text, check whether tex_dir and text_dir exist, respectively -- and if they don't, create them.
- Remove `TextWithBackground` from `text_mobject.py`. This was a non-working feature that basically copied code from `Text`. Code like that makes manim extremely hard to maintain in the long run. (Actually, also currently.)


## Motivation
Making `Tex` and `Text` robust against not created directories is required to use them in a setting where there is no `Scene` to set everything up (e.g., when writing tests for them outside of a scene, see #476).


## Testing Status
Added doctests for testing initialization of `Tex` and `Text`.


## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

behackl commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -77,6 +77,14 @@ class Text(SVGMobject):
+    Check whether writing text works::
+
+        >>> Text("The horse does not eat cucumber salad.")
+        <manim.mobject.svg.text_mobject.Text object at ...>
```
Personally, I don't like testing for the default string representation of Python classes that much and I would prefer to introduce a lightweight `__repr__` for these kind of things. My suggestion would be something like

```py

def __repr__(self):
    return f'Text("{self.original_text}")'
```
(or maybe also without the quotes or parentheses... I'm open for suggestions) so that we would have
```suggestion
        Text("The horse does not eat cucumber salad.")
```
as output for the doctest.

Any opinions on `__repr__`?

nilaybhatia replied:
The changed one seems better to me

naveen521kk commented on manim/utils/tex_file_writing.py:
```diff
@@ -34,6 +34,10 @@ def generate_tex_file(expression, tex_template, source_type):
+    if not os.path.exists(tex_dir):
+        os.makedirs(tex_dir)
+
     result = os.path.join(file_writer_config["tex_dir"], tex_hash(output)) + ".tex"
```
```suggestion
    result = os.path.join(tex_dir, tex_hash(output)) + ".tex"
```

nilaybhatia commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -324,71 +336,6 @@ def text2svg(self):
 
 
-# Following class is just a Little implementation of upcomming feautures. Ignore it for now.
-class TextWithBackground(Text):
```
Is it safe to remove this? I think it has something to do with `Code` mobject. Does 'Code` mobject work fine after removing? If it does not cause problems, then it's fine if we remove it now and later add whatever upcoming feature it was going to implement.

behackl replied:
`TextWithBackground` was not used anywhere else within manim, and in the form it was implemented it just called `Text.__init__` and did nothing else.

behackl commented:
Okay, I've added some very simple `__repr__` methods to both `Text` and `SingleStringMathTex` (from which all other TeX classes inherit -- weird, but maybe a story for another day) and I've added a bunch of more doctests so that creating text and tex is always tested.

Review at will!

leotrs commented on manim/mobject/svg/tex_mobject.py:
```diff
@@ -61,6 +71,9 @@ def __init__(self, tex_string, **kwargs):
             self.organize_submobjects_left_to_right()
 
+    def __repr__(self):
+        return f"{type(self).__name__}({repr(self.tex_string)})"
```
Question (non-blocking): what is the difference between `type(self).__name__` and `self.__class__.__name__`?

behackl replied:
The documentation says that they are generally the same (https://docs.python.org/3/library/functions.html#type).

I usually try to avoid accessing underscore methods directly, but I am fine with changing if you prefer `self.__class__.__name__`.

leotrs replied:
Nah, that's fine I was just curious

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -149,6 +157,9 @@ def __init__(self, text, **config):
             self.scale(TEXT_MOB_SCALE_FACTOR)
 
+    def __repr__(self):
+        return f"Text({repr(self.original_text)})"
```
Question: why does this one hardcode `Text` while the previous one does `type(self).__name__`?

behackl replied:
I was lazy and only wrote one `__repr__` for the entire Tex class structure, while `Text` does not have subclasses (any more, at least). This is why I thought using `type(self).__name__` might be overkill.

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -291,8 +302,12 @@ def text2svg(self):
 
         if self.font == "":
             if NOT_SETTING_FONT_MSG != "":
```
```suggestion
            if NOT_SETTING_FONT_MSG:
```

huguesdevimeux commented:
I _know_ you're going to hate me for that but ...


Do you think you could write a test for that ?

behackl commented:
> I _know_ you're going to hate me for that but ...
> 
> Do you think you could write a test for that ?

It is, in fact, already tested with the doctests I have added. :-)

huguesdevimeux commented:
> > I _know_ you're going to hate me for that but ...
> > Do you think you could write a test for that ?
> 
> It is, in fact, already tested with the doctests I have added. :-)

Could you clarify this ? I don't understand

behackl commented:
> > > I _know_ you're going to hate me for that but ...
> > > Do you think you could write a test for that ?
> > 
> > 
> > It is, in fact, already tested with the doctests I have added. :-)
> 
> Could you clarify this ? I don't understand

@huguesdevimeux The creation of `Text` etc is tested in https://github.com/ManimCommunity/manim/pull/492/files#diff-cf9247b89e1b4debba20823314a969f8R86 -- without an existing text_dir etc, these tests would fail.

If you would like to have a more explicit test, let me know what precisely you would like to have tested. I'll probably have to adapt to the PR that enabled Pango anyways.

behackl commented:
There was a merge conflict in the `TextWithBackground` class (the inheritance changed from `Text` to `CairoText`) -- but I still argue that it should be removed. It is not functional, and if someone wants to add it at a later point, then they are welcome to do so.

I don't find the current co-existance of `PangoText` and `CairoText` with `Text` still defaulting to `CairoText` a particularly good idea, and I think we should have switched the default and logged information that in case of problems one should fall back to the old implementation, `CairoText` -- but this is probably a discussion for a different issue / PR.

Anyhow, the reason I bring this up is the representation string: currently, `CairoText('asdf')` and `Text('asdf')` both have `Text('asdf')` as their representation string (and I think this is okay, as `Text` is just a wrapper with the logged warning). In 6e50d943 I've added a `__repr__` to `PangoText` that lets `PangoText('asdf')` also have `Text('asdf')` as its representation string, to keep this sort of "symmetry".

huguesdevimeux commented:
> If you would like to have a more explicit test, let me know what precisely you would like to have tested. I'll probably have to adapt to the PR that enabled Pango anyways.

I hear that, but I do think that at some point we will need a test for the whole directory generation. 
So, it's fine here, a test will be implemented in the future

behackl commented:
I would appreciate if this could get a second approval so I can stop resolving merge conflicts over and over. :sweat_smile:

huguesdevimeux reviewed (approved):
SGTM

nilaybhatia reviewed (approved):
LGTM!

behackl commented:
Thank you!
